### DISCUSSION TITLE
Pull out the value from a value channel

### DISCUSSION THREAD

**Discussion by multimeric** [nextflow-io/nextflow#2369] [Q&A] [answered]
DSL2 Workflows can only accept parameters that are channels. However I would like to convert one of these channels into a single Groovy value, since it will be much easier to work with this way. For example, say I have a channel of numbers, and I want to filter that channel down to only values greater than another number. Using only channels this looks like:

```groovy
nextflow.enable.dsl=2
workflow filter_channel {
    take:
        numbers
        threshold
    main:
        numbers | 
            combine(threshold) |
            filter {
                def (value, cutoff) = it
                value >= cutoff
            } | map {
                it[0]
            } | view
}

workflow {
    filter_channel(
        Channel.from([1, 2, 3, 4]),
        Channel.value(2)
    )
}

```

If we could pull values out of channels, I could simply write:
```groovy
workflow filter_channel {
    take:
        numbers
        threshold
    main:
        threshold_val = threshold | to_val()
        numbers | filter { it > threshold_val} | view
}
```

**Comment by bentsherman** [comment, marked as answer]
Hi @multimeric , sorry that your question was never answered. I'd like to answer even if you already figured it out since then, because it's an important question that we get from a lot of people. And it's top-rated on the discussions right now 😄 

The short answer is that it is not possible to pull a value out of a channel. Once you wrap something in a channel, you can't take it back out. Your "channel only" solution is actually *the only* solution. The only way to use the values in a channel is with an operator (or a process / subworkflow, which is a "kind" of operator).

Now for a longer answer, and maybe some speculation.

Some folks may remember that in DSL1 you could call `.val` or `.getVal()` on a channel, but no longer in DSL2. Even then, I suspect it was supported only by accident and was never really "supposed" to be used. But there is a reason why it was supported in DSL1 but not in DSL2.

In DSL1, pipelines were "eagerly" executed -- Nextflow would parse your script, build the workflow DAG, and run it all in the same step of "running your script". In DSL2, pipelines are "lazily" executed, meaning that "running your script" only builds the workflow DAG, and only after that does Nextflow execute the workflow itself. Lazy execution is *the thing* that enables modularity -- importing modules from separate files, composing workflows of workflows of workflows, etc. It also makes something like `channel.val` impossible, because Nextflow doesn't know what the value is when it executes this code. It hasn't even started the pipeline yet!

Having said all that, I agree that your second example is much more convenient. Maybe Nextflow should be able to *just figure out* the dataflow relationship here. The fact that it can't is the reason why channel logic is so wacky -- we have to describe the explicit dataflow that we want. If Nextflow could understand the *implicit dataflow*, maybe then we could write workflows with regular programming constructs -- assignments, expressions, if statements, so on. That's an interesting idea for a DSL3 😄

**Reply by multimeric** [reply]
Thanks, it's super helpful to have a definitive answer! (even if it's a tiny bit disappointing)

**Reply by alexnater** [reply]
Just to clarify: the statement "DSL2 Workflows can only accept parameters that are channels" is no longer valid, right? I can perfectly well pass a value to a workflow and use it  as shown in the second example. Is this something that got changed in a recent version of Nextflow?

**Reply by bentsherman** [reply]
You can pass regular values to a workflow, but the value will be implicitly wrapped in a value channel.

**Reply by alexnater** [reply]
However, if I pass an integer to a workflow and do a `.getClass()` within the workflow, I get java.lang.Integer as compared to groovyx.gpars.dataflow.DataflowVariable that I get if I explicitly pass the same value as a value channel.

**Reply by bentsherman** [reply]
I see... it must not be wrapping values at the workflow level, because they will be wrapped by the processes anyway.

@multimeric so I guess you could do something like:

```groovy
filter_channel( Channel.from([1, 2, 3, 4]), 2 )
```

Which would be fine if the threshold is a pipeline param, but not if it's actually coming from a channel in your real use case.

**Reply by lishengting** [reply]
It’s really inconvenient! 
I’m currently using splitFastq(n) and I want to split the file into 50 parts by given n=number_of_reads/50. 
However, the number_of_reads is contained in the output channel of a process. 
So, is it totally impossible to make splitFastq do this job?

**Reply by alexnater** [reply]
@lishengting That should be fairly easy to implement. No problem whatsoever to use the read counts in an output process as input for a downstream process.

**Reply by lishengting** [reply]
@alexnater Really? But how?
I also thought it should be fairly easy, but I have tried all kinds of methods and none of them worked, unless I don’t use splitFastq() and instead write my own process.
Can you provide an example using splitFastq()? Thanks!

**Reply by bentsherman** [reply]
@lishengting in this case you must compute the number of reads with a Groovy function rather than a process. You can use the [Files and I/O](https://nextflow.io/docs/latest/script.html#files-and-i-o) methods to read the file and count the number of reads.

**Reply by lishengting** [reply]
@bentsherman o, no you can't. the file doesn't exist at first, so a Groovy function will read a empty file because it doesn't wait like a process.

**Reply by bentsherman** [reply]
I see, the file is also coming from a process. In that case, you basically need a plain splitFastq() function that does not depend on channels, so that you could do something like this:
```groovy
MAKE_FASTQ
  | flatMap { it ->
    n = countFastq(it)
    return splitFastq(it, n / 50)
  }
```

You can probably write these functions yourself by taking inspiration from the splitFastq operator, see [FastqSplitter](https://github.com/nextflow-io/nextflow/blob/master/modules/nextflow/src/main/groovy/nextflow/splitter/FastqSplitter.groovy)

**Reply by lishengting** [reply]
@bentsherman That makes sense. I'll try it. Thanks a lot!

**Reply by bentsherman** [reply]
@lishengting I was thinking about this today and remembered that countFastq and splitFastq are already available as regular functions: https://nextflow.io/docs/latest/script.html#splitting-and-counting-records

So the code example I gave you should just work out of the box. Sorry I didn't realize it earlier

**Reply by blex-max** [reply]
@bentsherman After some extended discussions about very much wanting regular programming constructs in workflows, I was pointed to this thread - may I ask did anything come of this as an "idea for dsl3"?

**Reply by bentsherman** [reply]
I think we spoke on Slack, but for anyone else who is pointed to this thread -- you can follow [this PR](https://github.com/nf-core/fetchngs/pull/309) for my latest proposal for how to implement this idea
